urllib3/connectionpool.py

```diff
@@ -592,13 +594,13 @@ def urlopen(self, method, url, body=None, headers=None, retries=None,
             release_conn = True
             raise
 
-        except (TimeoutError, HTTPException, SocketError, ConnectionError) as e:
+        except (TimeoutError, HTTPException, SocketError, ProtocolError) as e:
             # Discard the connection for these exceptions. It will be
             # be replaced during the next _get_conn() call.
             conn = conn and conn.close()
             release_conn = True
 
-            if isinstance(e, SocketError) and self.proxy:
+            if isinstance(e, (SocketError, NewConnectionError)) and self.proxy:
```

Comment:
`TimeoutError` is the parent. But yea, our exception hierarchy needs a big makeover with explicit semantics of the request state during failure. :/